backend/wmg/pipeline/summary_cubes/calculate_markers.py

```diff
@@ -0,0 +1,620 @@
+        If None, the snapshot will be fetched from AWS.
+
+    n_markers - int, optional, default 10
+        Number of top markers to return. If None, all marker genes with effect size > 0 are returned.
+
+    percentile - float, optional, default 0.15
+        The percentile of effect sizes to select as the representative effect size.
+
+    Returns
+    -------
+    Dictionary of marker genes as keys and a dictionary of p-value and effect size as values.
+    """
+    (
+        context_agg,
+        target_agg,
+        groups_context_uniq,
+        n_target,
+        n_context,
+        target_data_nnz,
+        genes,
+        groups_indices_context,
+        genes_indices_context,
+        genes_indices_target,
+    ) = _prepare_indices_and_metrics(target_filters, context_filters, corpus=corpus)
+
+    target_data_sum = np.zeros((1, len(genes)))
+    target_data_sum[0, genes_indices_target] = list(target_agg["sum"])
+    target_data_sumsq = np.zeros((1, len(genes)))
+    target_data_sumsq[0, genes_indices_target] = list(target_agg["sqsum"])
+
+    context_data_sum = np.zeros((len(groups_context_uniq), len(genes)))
+    context_data_sum[groups_indices_context, genes_indices_context] = list(context_agg["sum"])
+    context_data_sumsq = np.zeros((len(groups_context_uniq), len(genes)))
+    context_data_sumsq[groups_indices_context, genes_indices_context] = list(context_agg["sqsum"])
+
+    pvals, effects = _run_ttest(
+        target_data_sum, target_data_sumsq, n_target, context_data_sum, context_data_sumsq, n_context
+    )
+    return _post_process_stats(
+        genes,
+        pvals,
+        effects,
+        target_data_nnz,
+        test="ttest",
+        n_markers=n_markers,
+        percentile=percentile,
+    )
+
+
+def _run_binom(nnz_thr1, n1, nnz_thr2, n2):
+    """
+    Calculate binomial test statistics.
+
+    Arguments
+    ---------
+    nnz_thr1 - np.ndarray (1 x M)
+        Array of number of nonzero expressions greater than a threshold (1) for target pop for each gene
+    n1 - np.ndarray (1 x M)
+        Number of cells in target population for each gene
+    nnz_thr2 - np.ndarray (N x M)
+        Array of number of nonzero expressions greater than a threshold (1) for context pops for each gene
+    n2 - np.ndarray (N x M)
+        Number of cells in context populations for each gene
+
+    Returns
+    -------
+    pvals_adj - np.ndarray
+        adjusted p-values for each comparison for each gene
+    effects - np.ndarray
+        effect sizes for each comparison for each gene
+    """
+    with np.errstate(divide="ignore", invalid="ignore"):
+        p_query = n1 / (n1 + n2)
+        size = nnz_thr1 + nnz_thr2
+        pvals = stats.binom.sf(nnz_thr1 + 1, size + 2, p_query)
+        mean_n = (n1 + n2) / 2
+        pn1 = n1 / mean_n
+        pn2 = n2 / mean_n
+        effects = np.log2((nnz_thr1 + pn1) / (n1 + 2 * pn1)) - np.log2((nnz_thr2 + pn2) / (n2 + 2 * pn2))
+
+    pvals[effects < 0] = np.nan
```

Comment:
```suggestion
```
not needed, `stats.binom.sf` is a one-sided test

Reply:
And actually these are converted later on to 1e-300, right?

https://github.com/chanzuckerberg/single-cell-data-portal/blob/df0ef4715e54af30be2d782e5991aa35dc2befb3/backend/wmg/pipeline/summary_cubes/calculate_markers.py#L394